compiler/src/main/kotlin/walkers/symbol.kt

```diff
@@ -0,0 +1,60 @@
+package dk.aau.cs.d409f19.cellumata.walkers
+
+import dk.aau.cs.d409f19.cellumata.ast.*
+import java.lang.Exception
+
+class SymbolException(val ident: String) : Exception("\"$ident\" was used before it was declared")
+
+class ScopeCheckVisitor(val symbolTable: SymbolTable) : BaseASTVisitor() {
+    override fun visit(node: RootNode) {
+        // Extract all states, functions, and neighbourhoods names first
+        node.body.filter { it is StateDecl }.forEach { symbolTable.insertSymbol((it as StateDecl).ident, it) }
+        node.body.filter { it is FuncDecl }.forEach { symbolTable.insertSymbol((it as FuncDecl).ident, it) }
+        node.body.filter { it is NeighbourhoodDecl }.forEach { symbolTable.insertSymbol((it as NeighbourhoodDecl).ident, it) }
+
+        // Visit constant declaration first
+        node.body.filter { it is ConstDecl }.forEach { visit(it) }
+
+        // Visit the other declarations
+        node.body.filter { it !is ConstDecl }.forEach { visit(it) }
+    }
+
+    override fun visit(node: VarExpr) {
+        if (symbolTable.getSymbol(node.ident) == null) {
+            throw SymbolException(node.ident)
+        }
+    }
+
+    override fun visit(node: ConstDecl) {
+        super.visit(node)
+        symbolTable.insertSymbol(node.ident, node)
+    }
+
+    override fun visit(node: StateDecl) {
+        symbolTable.createScope()
```

Comment:
Does not check if constants or expressions with variables has been declared.

Reply:
I think we should discus what expressions are allowed in state colors.

Reply:
I looked into the grammar and found that colors currently are limited to integer literals, so we should fill issue regarding expression as colors and the limitations we probably gonna have to put on them.